File: lib/model/product_join.dart
class ProductJoin {
  int? p_seq;
  String? p_name;
  int? p_price;
  int? p_stock;
  String? p_image;
  // String? p_description;
  // String? created_at;

  int? kc_seq;
  String? kc_name;
  int? qc_seq;
  String? qc_name;
  int? cc_seq;
  String? cc_name;
  int? sc_seq;
  String? sc_name;
  int? gc_seq;
  String? gc_name;

  int? m_seq;
  String? m_name;
  String? m_phone;
  String? m_address;

  ProductJoin({
    this.p_seq,
    this.p_name,
    this.p_price,
    this.p_stock,
    this.p_image,

    // this.p_description,
    // this.created_at,
    this.kc_seq,
    this.kc_name,
    this.qc_seq,
    this.qc_name,
    this.cc_seq,
    this.cc_name,
    this.sc_seq,
    this.sc_name,
    this.gc_seq,
    this.gc_name,

    this.m_seq,
    this.m_name,
    this.m_phone,
    this.m_address,
  });

  factory ProductJoin.fromJson(Map<String, dynamic> json) {
    return ProductJoin(
      p_seq: json['p_seq'] as int?,
      p_name: json['p_name'] as String?,
      p_price: json['p_price'] as int?,
      p_stock: json['p_stock'] as int?,
      p_image: json['p_image'] as String?,

      //p_description: json['p_description'] as String?,
      // created_at: json['created_at'] as String?,
      kc_seq: json['kc_seq'] as int?,
      kc_name: json['kc_name'] as String?,
      qc_seq: json['qc_seq'] as int?,
      qc_name: json['qc_name'] as String?,
      cc_seq: json['cc_seq'] as int?,
      cc_name: json['cc_name'] as String?,
      sc_seq: json['sc_seq'] as int?,
      sc_name: json['sc_name'] as String?,
      gc_seq: json['gc_seq'] as int?,
      gc_name: json['gc_name'] as String?,

      m_seq: json['m_seq'] as int?,
      m_name: json['m_name'] as String?,
      m_phone: json['m_phone'] as String?,
      m_address: json['m_address'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'p_seq': p_seq,
      'p_name': p_name,
      'p_price': p_price,
      'p_stock': p_stock,
      'p_image': p_image,
      //'p_description': p_description,
      //'created_at': created_at,
      'kc_seq': kc_seq,
      'kc_name': kc_name,
      'qc_seq': qc_seq,
      'qc_name': qc_name,
      'cc_seq': cc_seq,
      'cc_name': cc_name,
      'sc_seq': sc_seq,
      'sc_name': sc_name,
      'gc_seq': gc_seq,
      'gc_name': gc_name,
      'm_seq': m_seq,
      'm_name': m_name,
      'm_phone': m_phone,
      'm_address': m_address,
    };
  }
}

File: fastapi/app/api/product_join.py
"""
Product 복합 쿼리 API
- Product 중심의 JOIN 쿼리들
- Product + 모든 카테고리 (kind, color, size, gender) + Maker

개별 실행: python product_join.py
"""

from fastapi import APIRouter, Query
from typing import Optional
from app.database.connection import connect_db

router = APIRouter()


# ============================================
# Product 목록 + 모든 카테고리 + Maker
# ============================================
@router.get("/with_categories")
async def get_products_with_categories(
    maker_seq: Optional[int] = Query(None, description="제조사 ID (없으면 전체)"),
    kind_seq: Optional[int] = Query(None, description="종류 카테고리 ID"),
    color_seq: Optional[int] = Query(None, description="색상 카테고리 ID"),
    size_seq: Optional[int] = Query(None, description="사이즈 카테고리 ID"),
    gender_seq: Optional[int] = Query(None, description="성별 카테고리 ID")
):
    """
    Product 목록 + 모든 카테고리 + Maker 정보
    JOIN: Product + 모든 카테고리 + Maker
    용도: 제품 목록 화면 (필터링 가능)
    """
    conn = connect_db()
    curs = conn.cursor()
    
    try:
        # 동적 WHERE 절 구성
        conditions = []
        params = []
        
        if maker_seq:
            conditions.append("p.m_seq = %s")
            params.append(maker_seq)
        if kind_seq:
            conditions.append("p.kc_seq = %s")
            params.append(kind_seq)
        if color_seq:
            conditions.append("p.cc_seq = %s")
            params.append(color_seq)
        if size_seq:
            conditions.append("p.sc_seq = %s")
            params.append(size_seq)
        if gender_seq:
            conditions.append("p.gc_seq = %s")
            params.append(gender_seq)
        
        where_clause = "WHERE " + " AND ".join(conditions) if conditions else ""
        
        sql = f"""
        SELECT 
            p.p_seq,
            p.p_name,
            p.p_price,
            p.p_stock,
            p.p_image,
            kc.kc_name,
            cc.cc_name,
            sc.sc_name,
            gc.gc_name,
            m.m_name
        FROM product p
        JOIN kind_category kc ON p.kc_seq = kc.kc_seq
        JOIN color_category cc ON p.cc_seq = cc.cc_seq
        JOIN size_category sc ON p.sc_seq = sc.sc_seq
        JOIN gender_category gc ON p.gc_seq = gc.gc_seq
        JOIN maker m ON p.m_seq = m.m_seq
        {where_clause}
        ORDER BY p.p_seq DESC
        """
        
        curs.execute(sql, tuple(params))
        rows = curs.fetchall()
        
        result = [{
            'p_seq': row[0],
            'p_name': row[1],
            'p_price': row[2],
            'p_stock': row[3],
            'p_image': row[4],
            'kind_name': row[5],
            'color_name': row[6],
            'size_name': row[7],
            'gender_name': row[8],
            'maker_name': row[9]
        } for row in rows]
        
        return {"results": result}
    except Exception as e:
        return {"result": "Error", "errorMsg": str(e)}
    finally:
        conn.close()


# ============================================
# 개별 실행용 (테스트)


# ============================================
# 변경 이력
# ============================================
# 2026-01-01: 
#   - 라우터 경로 수정: prefix(/api/products)를 고려하여 /products/...를 제거
#     - /products/{product_seq}/full_detail → /{product_seq}/full_detail
#     - /products/with_categories → /with_categories
#     - /products/by_maker/{maker_seq}/with_categories → /by_maker/{maker_seq}/with_categories
#     - /products/by_category → /by_category